### PR TITLE
feat(router): enable `net.src.*` and `net.dst.*` in http expression

### PR DESCRIPTION
**Pull request by chronolaw**
### Summary

KAG-2963
KAG-3032

### Checklist

- [x] The Pull Request has tests
- [x] A changelog file has been created under `changelog/unreleased/kong` or `skip-changelog` label added on PR if changelog is unnecessary. [README.md](https://github.com/Kong/gateway-changelog/README.md)
- [ ] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE

### Full changelog

* [Implement ...]

### Issue reference

Fix #_[issue number]_

### CODE REVIEW

File: kong/router/atc.lua [hunk 1]
```diff
@@ -67,7 +67,10 @@ do
                   "http.queries.*",
                  },
 
-    ["Int"]    = {"net.port",
+    ["Int"]    = {"net.src.port", "net.dst.port",
+                 },
+
+    ["IpAddr"] = {"net.src.ip", "net.dst.ip",
```
Comment (mikefero) [inline kong/router/atc.lua:73]: What kind of impact will this have on version compatibility in hybrid mode for older data planes?

Comment (chronolaw) [reply kong/router/atc.lua:73]: This PR did not change the expression stored in db, only convert it adhoc, so if dp does not upgrade to the latest version, it will still work well.

File: kong/router/expressions.lua [hunk 2]
```diff
@@ -19,8 +19,28 @@ local PROTOCOLS_OVERRIDE = {
...
-local function get_exp_and_priority(route)
+-- net.port => net.dst.port
+local function verify_expression(route)
   local exp = route.expression
+
+  if not exp then
+    return nil
+  end
+
+  if exp:find("net.port", 1, true) then
```
Comment (StarlightIbuki) [inline kong/router/expressions.lua:30]: This is not robust enough. For example:

```
http.header.host ="net.port.com"
```

(Just to illustrate.)

File: kong/router/expressions.lua [hunk 3]
```diff
@@ -19,8 +19,28 @@ local PROTOCOLS_OVERRIDE = {
...
+    ngx.log(ngx.WARN, "The field 'net.port' of expression is deprecated, " ..
+                      "please use 'net.dst.port' instead.")
+
+    return exp:gsub("net%.port", "net.dst.port")
```
Comment (StarlightIbuki) [inline kong/router/expressions.lua:34]: Could we use the returned value to determine if `net.port` is used?

File: kong/router/expressions.lua (hunk 2 above)
Comment (StarlightIbuki) [reply kong/router/expressions.lua:30]: I prefer to do this on the syntax level rather than a string replacing.

File: spec/01-unit/01-db/01-schema/06-routes_spec.lua [hunk 4]
```diff
@@ -1510,4 +1510,43 @@ describe("routes schema (flavor = expressions)", function()
     -- verified by `schema/typedefs.lua`
     assert.truthy(errs["@entity"])
   end)
+
+  it("http route still supports net.port but with warning", function()
```
Comment (StarlightIbuki) [inline spec/01-unit/01-db/01-schema/06-routes_spec.lua:1514]: We should also make assertions on the warning messages.

File: kong/router/atc.lua [hunk 5]
```diff
@@ -666,14 +673,38 @@ function _M:exec(ctx)
...
+
   -- cache lookup
 
   local cache_key = (req_method  or "") .. "|" ..
                     (req_uri     or "") .. "|" ..
                     (req_host    or "") .. "|" ..
                     (sni         or "") .. "|" ..
                     (headers_key or "") .. "|" ..
-                    (queries_key or "")
+                    (queries_key or "") .. "|" ..
```
Comment (StarlightIbuki) [inline kong/router/atc.lua:702]: Where do we test this change?

**Review by StarlightIbuki — CHANGES_REQUESTED** [review]
I feel this deprecation is more properly done from the ATC router side. What do you think?

**Comment by StarlightIbuki** [comment]
Also, we should split this into 2 PRs, to let the title be precise.

**Comment by chronolaw** [comment]
> I feel this deprecation is more properly done from the ATC router side. What do you think?

No, atc-router library has no schema definition, we only define schema here.

**Comment by chronolaw** [comment]
> Also, we should split this into 2 PRs, to let the title be precise.

Perhaps we can not do this, add `net.src.*` and `net.dst.*` will cause the change of cache key at the same time.

File: spec/01-unit/01-db/01-schema/06-routes_spec.lua (hunk 4 above)
Comment (chronolaw) [reply spec/01-unit/01-db/01-schema/06-routes_spec.lua:1514]: Could you give me an example about how to check the warning message? thanks.

Comment (StarlightIbuki) [reply spec/01-unit/01-db/01-schema/06-routes_spec.lua:1514]: For integrated tests we use `assert.logfile`; for unit tests we use [`assert.spy`](https://lunarmodules.github.io/busted/).

File: spec/01-unit/01-db/01-schema/06-routes_spec.lua [hunk 6]
```diff
@@ -1512,6 +1512,7 @@ describe("routes schema (flavor = expressions)", function()
   end)
 
   it("http route still supports net.port but with warning", function()
+    local log = spy.on(ngx, "log")
```
Comment (StarlightIbuki) [inline spec/01-unit/01-db/01-schema/06-routes_spec.lua:1515]: Remember to recover the `ngx.log` after the test. Use setup and teardown to make sure it won't affect other tests.

Comment (chronolaw) [reply spec/01-unit/01-db/01-schema/06-routes_spec.lua:1516]: Please check the `finally` to ensure I'm corret.

Comment (chronolaw) [reply spec/01-unit/01-db/01-schema/06-routes_spec.lua:1516]: I got an error `spec/01-unit/01-db/01-schema/06-routes_spec.lua:1519:7: setting read-only field log of global ngx`,
What is the right way to this?

File: spec/01-unit/01-db/01-schema/06-routes_spec.lua [hunk 7]
```diff
@@ -1512,7 +1512,13 @@ describe("routes schema (flavor = expressions)", function()
   end)
 
   it("http route still supports net.port but with warning", function()
+    local ngx_log = ngx.log
     local log = spy.on(ngx, "log")
+
+    finally(function()
+      ngx.log = ngx_log
```
Comment (StarlightIbuki) [inline spec/01-unit/01-db/01-schema/06-routes_spec.lua]:
```suggestion
      ngx.log = ngx_log -- luacheck: ignore
```

Do this to avoid warning.

Comment (chronolaw) [reply spec/01-unit/01-db/01-schema/06-routes_spec.lua]: done.

File: kong/router/expressions.lua (hunk 2 above)
Comment (chronolaw) [reply kong/router/expressions.lua]: I think that it should be a corner case.

File: kong/router/expressions.lua (hunk 3 above)
Comment (chronolaw) [reply kong/router/expressions.lua]: changed the code.

File: kong/router/atc.lua (hunk 5 above)
Comment (chronolaw) [reply kong/router/atc.lua:702]: It seems that we have no test case for cache key.

File: kong/router/expressions.lua (hunk 2 above)
Comment (StarlightIbuki) [reply kong/router/expressions.lua]: > I think that it should be a corner case.

And we need to cover corner cases.

File: kong/router/atc.lua (hunk 1 above)
Comment (dndx) [reply kong/router/atc.lua:73]: @mikefero Will send you a PM

**Comment by dndx** [comment]
@StarlightIbuki Please review again.

We need a temporary solution for deprecation. Considering the limited use case of expressions, I doubt we need to make the migration that fancy. I'm okay with a solution that doesn't understand the syntax.

File: kong/router/expressions.lua [hunk 8]
```diff
@@ -19,8 +24,34 @@ local PROTOCOLS_OVERRIDE = {
...
+    return nil
+  end
+
+  if not exp:find("net.port", 1, true) then
+    return exp
+  end
+
+  -- there is "net.port" in expression
+
+  local new_exp = re_gsub(exp, NET_PORT_REG, NET_PORT_REPLACE)
```
Comment (StarlightIbuki) [inline kong/router/expressions.lua:41]:
```suggestion
  local new_exp = re_gsub(exp, NET_PORT_REG, NET_PORT_REPLACE, "joi")
```

File: kong/router/expressions.lua [hunk 9]
```diff
@@ -1,11 +1,16 @@
...
 
+local re_gsub = ngx.re.gsub
+
+
 local atc = require("kong.router.atc")
 local gen_for_field = atc.gen_for_field
 
 
 local OP_EQUAL    = "=="
+local NET_PORT_REG = [[(net\.port)(\s*)([=><!])]]
```
Comment (StarlightIbuki) [inline kong/router/expressions.lua:12]: Good enough for me now.

**Review by StarlightIbuki — APPROVED** [review]
A minor suggestion on performance. Otherwise good enough.

File: kong/router/expressions.lua (hunk 8 above)
Comment (chronolaw) [reply kong/router/expressions.lua:41]: According to ngx_lua's doc, we should not use 'o' option in `ngx.re.gsub`. (https://github.com/openresty/lua-nginx-module#lua_regex_cache_max_entries).

And I think that it will not be a hot path.

File: kong/router/expressions.lua (hunk 9 above)
Comment (chronolaw) [reply kong/router/expressions.lua:12]: thanks.

File: kong/router/expressions.lua (hunk 8 above)
Comment (dndx) [reply kong/router/expressions.lua:41]: @chronolaw I think we should, the doc says:

> Do not activate the o option for regular expressions (and/or replace string arguments for [ngx.re.sub](https://github.com/openresty/lua-nginx-module#ngxresub) and [ngx.re.gsub](https://github.com/openresty/lua-nginx-module#ngxregsub)) **that are generated on the fly**...

`NET_PORT_REG` is clearly not generated on the fly, so `o` would be appropriate here.

Comment (dndx) [reply kong/router/expressions.lua:41]: Also, `jo` would be fine, we want to be case-sensitive (no `i`).

Comment (chronolaw) [reply kong/router/expressions.lua]: done, adding "jo"

**Comment by StarlightIbuki** [comment]
@chronolaw What is the status of this PR?

**Comment by chronolaw** [comment]
> @chronolaw What is the status of this PR?

After a offline meeting we decide to re-work this feature.

File: kong/router/expressions.lua [hunk 10]
```diff
@@ -19,8 +24,34 @@ local PROTOCOLS_OVERRIDE = {
 }
 
 
-local function get_exp_and_priority(route)
+-- net.port => net.dst.port
+local function verify_expression(route)
```
Comment (dndx) [inline kong/router/expressions.lua:28]:
```suggestion
local function transform_expression(route)
```

Might be better. Verify should only return `true` or `false`.

File: kong/router/expressions.lua [hunk 11]
```diff
@@ -19,8 +24,34 @@ local PROTOCOLS_OVERRIDE = {
...
+  if not exp:find("net.port", 1, true) then
+    return exp
+  end
+
+  -- there is "net.port" in expression
+
+  local new_exp = re_gsub(exp, NET_PORT_REG, NET_PORT_REPLACE, "jo")
+
+  if exp ~= new_exp then
+    ngx.log(ngx.WARN, "The field 'net.port' of expression is deprecated, " ..
```
Comment (dndx) [inline kong/router/expressions.lua:44]:
```suggestion
    ngx.log(ngx.WARN, "The field 'net.port' of expression is deprecated and will be removed in the upcoming major release, " ..
```

**Review by ADD-SP — CHANGES_REQUESTED** [review]

File: kong/router/expressions.lua (hunk 9 above)
[the remark this answers is not in the document]
Comment (ADD-SP) [reply kong/router/expressions.lua:12]: This regex doesn't cause any bug now, but it might become a bug with more and more String-type fields are added.

Is it possible to add a regression test to prevent that?

File: kong/router/expressions.lua [hunk 12]
```diff
@@ -19,8 +24,35 @@ local PROTOCOLS_OVERRIDE = {
 }
 
 
-local function get_exp_and_priority(route)
+-- net.port => net.dst.port
+local function transform_expression(route)
```
[the remark this answers is not in the document]
Comment (chronolaw) [reply kong/router/expressions.lua:28]: This function may have other purposes in the future, let's keep it.

File: kong/router/expressions.lua (hunk 9 above)
[the remark this answers is not in the document]
Comment (chronolaw) [reply kong/router/expressions.lua:12]: `net.port` is a undocumented feature, I think that now there is few people using it, and in next major version we will remove this field.

File: changelog/unreleased/kong/deprecate_net_port_field_in_expression.yml
Comment (ADD-SP) [inline changelog/unreleased/kong/deprecate_net_port_field_in_expression.yml:1]: The `net.port` is an undocumented field, Should we mention it in the changelog?

File: kong/router/expressions.lua [hunk 13]
```diff
@@ -19,8 +24,35 @@ local PROTOCOLS_OVERRIDE = {
...
+  end
+
+  -- there is "net.port" in expression
+
+  local new_exp = re_gsub(exp, NET_PORT_REG, NET_PORT_REPLACE, "jo")
+
+  if exp ~= new_exp then
+    ngx.log(ngx.WARN, "The field 'net.port' of expression is deprecated " ..
+                      "and will be removed in the upcoming major release, " ..
+                      "please use 'net.dst.port' instead.")
```
Comment (ADD-SP) [inline kong/router/expressions.lua:46]: The `net.port` is an undocumented feature, so it might be used by our other products, if the expression is not written by the customer directly (such as KIC), this message might confuse them.

We should consider the message carefully.

**Review by ADD-SP — APPROVED** [review]

File: changelog/unreleased/kong/support_net_src_dst_field_in_expression.yml [hunk 14]
```diff
@@ -0,0 +1,4 @@
+message: |
+  make `net.src.*` and `net.dst.*` accessible in http routes as well.
```
Comment (dndx) [inline changelog/unreleased/kong/support_net_src_dst_field_in_expression.yml:2]:
```suggestion
  `net.src.*` and `net.dst.*` match fields are now accessible in HTTP routes defined using expressions.
```

**Review by dndx — APPROVED** [review]

**Comment by flrgh** [comment]
Is this ready to merge, or are we waiting for any reason?
